File: problems/1732/1732.find-the-highest-altitude.py
#
# @lc app=leetcode id=1732 lang=python3
#
# [1732] Find the Highest Altitude
#


# @lc code=start
class Solution:
    def largestAltitude(self, gain: List[int]) -> int:
        # n + 1 points at different altitudes
        # starts from 0
        # prefix sum

        curr_max = 0
        curr = 0
        for val in gain:
            curr += val
            curr_max = max(curr_max, curr)

        return curr_max


# @lc code=end

File: problems/1732/1732.find-the-highest-altitude.java
/*
 * @lc app=leetcode id=1732 lang=java
 *
 * [1732] Find the Highest Altitude
 */

// @lc code=start
class Solution {
    public int largestAltitude(int[] gain) {
        /*
         * highest altitude
         *
         * intuition:
         * - iterate and keep track of max altitude
         *
         * O(N) time, O(1) space
         */

        int ans = 0;
        int prev = 0;
        for (int val : gain) {
            prev += val;
            if (prev > ans) {
                ans = prev;
            }
        }

        return ans;
    }
}
// @lc code=end
